File: static/java/Formulas.java
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/** Top-level formula combinators matching Airtable's logical functions. */
public final class Formulas {

  private Formulas() {}

  /** Combine formulas with AND logic. Empty strings are filtered out. */
  public static String and(String... args) {
    return and(Arrays.asList(args));
  }

  public static String and(List<String> args) {
    List<String> filtered = args.stream().filter(a -> !a.isEmpty()).toList();
    return switch (filtered.size()) {
      case 0 -> "";
      case 1 -> filtered.get(0);
      default -> "AND(" + String.join(",", filtered) + ")";
    };
  }

  /** Combine formulas with OR logic. Empty strings are filtered out. */
  public static String or(String... args) {
    return or(Arrays.asList(args));
  }

  public static String or(List<String> args) {
    List<String> filtered = args.stream().filter(a -> !a.isEmpty()).toList();
    return switch (filtered.size()) {
      case 0 -> "";
      case 1 -> filtered.get(0);
      default -> "OR(" + String.join(",", filtered) + ")";
    };
  }

  /** Negate a formula. */
  public static String not(String formula) {
    return "NOT(" + formula + ")";
  }

  /** Exclusive OR. Empty strings are filtered out. */
  public static String xor(String... args) {
    return xor(Arrays.asList(args));
  }

  public static String xor(List<String> args) {
    List<String> filtered = args.stream().filter(a -> !a.isEmpty()).toList();
    return switch (filtered.size()) {
      case 0 -> "";
      case 1 -> filtered.get(0);
      default -> "XOR(" + String.join(",", filtered) + ")";
    };
  }

  // ---- shared escaping / wrapping helpers (package-private) ----

  /**
   * Escape a value for an Airtable double-quoted formula string. Backslashes are escaped FIRST,
   * then quotes — otherwise a trailing {@code \} leaves the closing quote live and the remainder of
   * the value executes as formula code.
   */
  static String escapeFormulaString(String value) {
    return value.replace("\\", "\\\\").replace("\"", "\\\"");
  }

  /** Escape a value for an Airtable single-quoted formula string. */
  static String escapeFormulaStringSingle(String value) {
    return value.replace("\\", "\\\\").replace("'", "\\'");
  }

  static String wrapField(String field, boolean caseSensitive, boolean trim) {
    String v = field;
    if (trim) {
      v = "TRIM(" + v + ")";
    }
    if (!caseSensitive) {
      v = "LOWER(" + v + ")";
    }
    return v;
  }

  static String wrapValueStr(String value, boolean caseSensitive, boolean trim) {
    String v = "\"" + escapeFormulaString(value) + "\"";
    if (trim) {
      v = "TRIM(" + v + ")";
    }
    if (!caseSensitive) {
      v = "LOWER(" + v + ")";
    }
    return v;
  }

  static String joinMapped(List<String> values, java.util.function.Function<String, String> fn) {
    return values.stream().map(fn).collect(Collectors.joining(","));
  }
}

File: static/csharp/Formulas.cs
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace MyAirtable;

/// <summary>Top-level formula combinators matching Airtable's logical functions.</summary>
public static class Formulas
{
    /// <summary>Combine formulas with AND logic. Empty strings are filtered out.</summary>
    public static string And(params string[] args) => And((IEnumerable<string>)args);

    public static string And(IEnumerable<string> args) => Combine("AND", args);

    /// <summary>Combine formulas with OR logic. Empty strings are filtered out.</summary>
    public static string Or(params string[] args) => Or((IEnumerable<string>)args);

    public static string Or(IEnumerable<string> args) => Combine("OR", args);

    /// <summary>Exclusive OR. Empty strings are filtered out.</summary>
    public static string Xor(params string[] args) => Xor((IEnumerable<string>)args);

    public static string Xor(IEnumerable<string> args) => Combine("XOR", args);

    /// <summary>Negate a formula.</summary>
    public static string Not(string formula) => "NOT(" + formula + ")";

    private static string Combine(string fn, IEnumerable<string> args)
    {
        var filtered = args.Where(a => !string.IsNullOrEmpty(a)).ToList();
        return filtered.Count switch
        {
            0 => "",
            1 => filtered[0],
            _ => $"{fn}({string.Join(",", filtered)})",
        };
    }

    // ---- shared escaping / wrapping helpers (internal) ----

    /// <summary>
    /// Escape a value for an Airtable double-quoted formula string. Backslashes are escaped FIRST,
    /// then quotes — otherwise a trailing <c>\</c> leaves the closing quote live and the remainder
    /// of the value executes as formula code.
    /// </summary>
    internal static string EscapeFormulaString(string value) =>
        value.Replace("\\", "\\\\").Replace("\"", "\\\"");

    /// <summary>Escape a value for an Airtable single-quoted formula string.</summary>
    internal static string EscapeFormulaStringSingle(string value) =>
        value.Replace("\\", "\\\\").Replace("'", "\\'");

    /// <summary>Format a number for a formula (invariant culture — no locale separators).</summary>
    internal static string Num(double value) => value.ToString(CultureInfo.InvariantCulture);

    internal static string WrapField(string field, bool caseSensitive, bool trim)
    {
        var v = field;
        if (trim)
            v = "TRIM(" + v + ")";
        if (!caseSensitive)
            v = "LOWER(" + v + ")";
        return v;
    }

    internal static string WrapValueStr(string value, bool caseSensitive, bool trim)
    {
        var v = "\"" + EscapeFormulaString(value) + "\"";
        if (trim)
            v = "TRIM(" + v + ")";
        if (!caseSensitive)
            v = "LOWER(" + v + ")";
        return v;
    }
}
